### PR TITLE
feat(matchBaseBranches): Support regex expression

### LINKED ISSUE #13320: PackageRule option `matchBaseBranches` should support regex wildcard
### What would you like Renovate to be able to do?

I want to configure special package rules only for all of my release branches., but not for main branch.
Now I use matchBaseBranches to list all of release branches, but I have to update this for every new release branch created.

**Current configuration**
```json
"matchBaseBranches": [
    "release/5.8.x",
    "release/5.7.x",
    "release/5.6.x"
],
```

The reason for special rules for all the release branches to disable Major upgrades and other stuff.

### If you have any ideas on how this should be implemented, please tell us here.

**Desired configuration**
```json
"matchBaseBranches": [
    "/^release\\//"
],
```
_Update: Update desired configuration after suggestion of rarkins_

### Is this a feature you are interested in implementing yourself?

Yes

### PR DESCRIPTION
**Pull request by marcovmun**
## Changes:

Support regex expression for configuration option `matchBaseBranches` (#13320)

## Context:

Closes #13320

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
needs CLA signing

File: lib/config/options/index.ts [hunk 1]
```diff
@@ -852,7 +852,7 @@ const options: RenovateOptions[] = [
   {
     name: 'matchBaseBranches',
     description:
-      'List of branches to match (e.g. ["master"]). Valid only within `packageRules` object.',
+      'List of branches to match (e.g. ["master"]) or list of regex patterns (e.g. ["^/release\\/.*/"]). Valid only within `packageRules` object.',
```
Comment (rarkins) [inline lib/config/options/index.ts:855]:
```suggestion
      'List of branches to match (e.g. ["master"]) or list of regex patterns (e.g. ["/^release\\/.*/"]). Valid only within `packageRules` object.',
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by marcovmun** [comment]
I needed to force push the pull request to change the email address linked to the old commit.
Also fixed the documentation mistake.

File: lib/config/options/index.ts (hunk 1 above)
Comment (marcovmun) [reply lib/config/options/index.ts]: Changed in new commit with correct email address for CLA

File: lib/util/package-rules.ts [hunk 2]
```diff
@@ -100,11 +100,21 @@ function matchesRule(
     positiveMatch = true;
   }
   if (matchBaseBranches.length) {
-    const isMatch = matchBaseBranches.includes(baseBranch);
-    if (!isMatch) {
+    const hasBaseBranchMatch = matchBaseBranches.some(
```
Comment (viceice) [inline lib/util/package-rules.ts:103]:
```suggestion
    const isMatch = matchBaseBranches.some(
```

File: lib/util/package-rules.ts [hunk 3]
```diff
@@ -100,11 +100,21 @@ function matchesRule(
...
+        if (isAllowedPred) {
+          return isAllowedPred(baseBranch);
+        }
+        return matchBaseBranch === baseBranch;
+      }
+    );
+
+    if (hasBaseBranchMatch) {
+      positiveMatch = true;
+    } else {
```
Comment (viceice) [inline lib/util/package-rules.ts:115]:
```suggestion
    if (!isMatch) {
```

File: lib/util/package-rules.ts [hunk 4]
```diff
@@ -100,11 +100,21 @@ function matchesRule(
...
       return false;
     }
-    positiveMatch = true;
```
Comment (viceice) [inline lib/util/package-rules.ts:107]: revert this

**Review by viceice — COMMENTED** [review]
some small things to minimize changes

File: lib/config/options/index.ts [hunk 5]
```diff
@@ -852,7 +852,7 @@ const options: RenovateOptions[] = [
   {
     name: 'matchBaseBranches',
     description:
-      'List of branches to match (e.g. ["master"]). Valid only within `packageRules` object.',
+      'List of branches to match (e.g. ["master"]) or list of regex patterns (e.g. ["/^release\\/.*/"]). Valid only within `packageRules` object.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:855]:
```suggestion
      'List of branches to match (e.g. `["master"]`) or list of regex patterns (e.g. `["/^release\\/.*/"]`). Valid only within `packageRules` object.',
```

This makes the final text look like this, which should be easier to read. 😄 

> List of branches to match (e.g. `["master"]`) or list of regex patterns (e.g. `["/^release\\/.*/"]`). Valid only within `packageRules` object.

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -1489,6 +1489,8 @@ Use this field to restrict rules to a particular branch. e.g.
 }
 ```
 
+This field also supports Regular Expressions if they begin and end with `/`.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1492]: Do we want to add an example like this one below this line?
I've grabbed the regex from the `description` field in the `lib/config/options/index.ts` file in this PR.

```json
{
  "packageRules": [
    {
      "matchBaseBranches": ["main"],
      "excludePackagePatterns": ["/^release\\/.*/"],
      "enabled": false
    }
  ]
}
```

File: lib/config/options/index.ts (hunk 5 above)
Comment (viceice) [reply lib/config/options/index.ts:855]: They can also be mixed, so it should be a list of branches or regex patterns

Comment (HonkingGoose) [reply lib/config/options/index.ts:855]: > They can also be mixed, so it should be a list of branches or regex patterns

Can you suggest a better `description`? I'm not sure what you want me to change... 🙈

Comment (marcovmun) [reply lib/config/options/index.ts:855]: Maybe something linke this?
```
List of strings containing exact matches (e.g. `["master"]`) and/or regex expressions (e.g. `["/^release\\/.*/"]`). Valid only within `packageRules` object.
```

File: docs/usage/configuration-options.md (hunk 6 above)
Comment (marcovmun) [reply docs/usage/configuration-options.md:1492]: I had explicitly not done this, to not make documentation unnecessarily long. Bus yes I can add this. 
I can also make one documentatie block with both options:
```
Use this field to restrict rules to a particular branch. This field also supports Regular Expressions if they begin and end with `/`. e.g.

```json
{
  "packageRules": [
    {
      "matchBaseBranches": [
             "main"
             "/^release\\/.*/"
      ],
      "excludePackagePatterns": ["^eslint"],
      "enabled": false
    }
  ]
}
```
```

File: lib/config/options/index.ts (hunk 5 above)
Comment (HonkingGoose) [reply lib/config/options/index.ts:855]: I like it! 😄

File: docs/usage/configuration-options.md (hunk 6 above)
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:1492]: I think having it in one block can be confusing.
JSON doesn't allow comments, so you cannot put a comment behind each example saying "this is regex" and " this is normal stuff".

I would like 2 separate examples, even if it is a bit longer. 😄

**Review by viceice — DISMISSED** [review]

File: lib/config/options/index.ts [hunk 7]
```diff
@@ -852,7 +852,7 @@ const options: RenovateOptions[] = [
   {
     name: 'matchBaseBranches',
     description:
-      'List of branches to match (e.g. ["master"]). Valid only within `packageRules` object.',
+      'List of strings containing exact matches (e.g. `["master"]`) and/or regex expressions (e.g. `["/^release\\/.*/"]`). Valid only within `packageRules` object.',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:855]:
```suggestion
      'List of strings containing exact matches (e.g. `["main"]`) and/or regex expressions (e.g. `["/^release\\/.*/"]`). Valid only within `packageRules` object.',
```

I'd rather use `main` here, now that many Git hosting providers have switched to this name for the default branch. 😉

Comment (marcovmun) [reply lib/config/options/index.ts]: Of course, stupid of me. I'm so used to master yet. I changed it

Comment (HonkingGoose) [reply lib/config/options/index.ts]: Thank you for applying the suggestion! 😄 

Don't feel bad, the reference to `master` was in there before you touched the file, I should have updated this sooner. 🙈

**Review by HonkingGoose — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**